Use PEM certificates loaded from secrets for KafkaConnect

tinaselenge opened the pull request:
### Type of change

_Select the type of your PR_

- Refactoring

### Description

- TLS certificates are loaded from secrets directly using KubernetesSecretConfigProvider, therefore also removes the population of the certificate related environment variables and the script for preparing TLS. 
- Use key pattern (*.crt) of KubernetesSecretConfigProvider to load multiple certificates from a single secret for `ssl.truststore.certificates` configuration.  
- OAuth truststore is however configured differently, because multiline line certificates in Jaas config is not parsed correctly. Instead it will copy all the trusted certificates in PEM format into an internal secret under a single key and then volume mount it  to use for `ssl.truststore.location` of oauth configuration. 
- Add RBAC right for reading secrets for KafkaConnect service account.


Resolves part of https://github.com/strimzi/strimzi-kafka-operator/issues/11294

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

tinaselenge commented:
@scholzj @ppatierno @katheris can one of you please kick off the regression tests? I ran some of the relevant ST locally which seemed to pass, but there are many I haven't run so would like to try running the full suite. Thanks!

tinaselenge commented:
thanks @scholzj for kicking off the tests. Looks like the failing tests are not related to this PR (time out failures in CruiseControlST which doesn't deploy any connect resource). 

I will mark this PR ready for review now :) .

scholzj commented on CHANGELOG.md:
```diff
@@ -12,6 +12,7 @@
 * Dependency updates (Vert.x 4.5.12, Netty 4.1.117.Final)
 * Moved Kafka Connect configuration to the ConfigMap created by the operator.
 * Update Kafka Exporter to [1.9.0](https://github.com/danielqsj/kafka_exporter/releases/tag/v1.9.0)
+* Kafka Connect SSL configurations now uses Kafka Kubernetes Config Provider to load certificates directly from secrets.
```
You should add a note to the `Major changes, deprecations and removals` section as well -> a warning to any users overriding the keystore / truststore location in their Connector configurations that they will need to update them to override the PEM files.

scholzj reviewed (commented):
I left some nits, but otherwise it mostly looks good.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilder.java:
```diff
@@ -84,18 +89,18 @@ public KafkaConnectConfigurationBuilder withTls(ClientTls tls) {
-                writer.println("ssl.truststore.location=/tmp/kafka/cluster.truststore.p12");
-                writer.println("ssl.truststore.password=" + PLACEHOLDER_CERT_STORE_PASSWORD_CONFIG_PROVIDER_ENV_VAR);
-                writer.println("ssl.truststore.type=PKCS12");
+                String configProviderValue = String.format(PLACEHOLDER_SECRET_TEMPLATE_KUBE_CONFIG_PROVIDER, reconciliation.namespace(), tls.getTrustedCertificates().get(0).getSecretName(), "*.crt");
```
Fair point. There can be multiple secrets. Maybe you should copy them into a single internal Secret and use that in the config provider? I think the config providers have anyway issues when the files/secrets get missing. So having it stored internally under our control might be better?

fvaleri replied:
> having it stored internally under our control might be better?

This does not scale well. In the worst case, we could potentially have many Secrets containing many certificates each that won't fit in a single "internal" Secret. Looking at our examples, it would also mean duplicating my-cluster-cluster-ca-cert. Can't we add some validation and raise an error in case one or more configured Secrets are missing?

scholzj replied:
If any user has more trusted certificates than fit into a Secret they deserve to burn in hell.

The isolation from the user secrets is really important and cannot be replaced with validation. We saw in the past that Connect does not handle things like the secrets disappearing well.

tinaselenge replied:
I was wondering how common it is in general, users having multiple secrets for trusted certificates. If it's not that common, I doubt there are many cases where so many secrets and certificates that would exceed a single secret are used. 

@scholzj I took your suggestion on the internal secret combining all the user secrets for certs. Please let me know, if it does or does not align with what you were suggesting.

tinaselenge replied:
I had another thought on this. Maybe we could just use multiple config provider templated values joined by a newline since that's what Kubernetes config provider is doing with a patterned key (e.g. `*.crt`). So that configuration would look like this:
```
ssl.truststore.certificates=${strimzisecrets:namespace/my-secret:*.crt}\n${strimzisecrets:namespace/my-another-secret:*.crt}\n
```
I think this is simpler than creating an internal secret.

tinaselenge replied:
I reverted the change for creating an internal secret (the [commit](https://github.com/tinaselenge/strimzi-kafka-operator/commit/528bb0581bed61ef5010a1f5b4df56f09b235b08) is another branch) and implemented the change I suggested above. Running ConnectST seems to work, but I haven't tried it with multiple trusted certificates from different secret sources.

fvaleri replied:
Very good, but the ST is always using one trusted cert. Maybe you can also improve that test to have two of them.

fvaleri commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilder.java:
```diff
@@ -26,22 +28,25 @@
+     * @param bootstrapServers Kafka cluster bootstrap servers to connect to
      */
-    public KafkaConnectConfigurationBuilder(String bootstrapServers) {
+    public KafkaConnectConfigurationBuilder(Reconciliation reconciliation, String bootstrapServers) {
```
We only need namespace from the Reconciliation object, which is already available as KafkaConnectCluster class variable from the AbstractModel, so we could directly pass that.

ppatierno replied:
I think we have this pattern in all the other builders. Passing just the `Reconciliation` object narrows the scope, while with the entire `KafkaConnectCluster` object (if this is what you mean) you are potentially giving access to more stuff we don't need here.

tinaselenge replied:
I have followed the pattern in the other builders but I think passing just the namespace would work here too.

ppatierno replied:
I would stick with what we have elsewhere so passing reconciliation object.

fvaleri replied:
> with the entire KafkaConnectCluster object (if this is what you mean)

This is not what I meant. I was suggesting to *only* use namespace. Anyway, it's ok to pass reconciliation for consistency.

ppatierno commented on CHANGELOG.md:
```diff
@@ -36,6 +37,8 @@
   Please use the template section to configure additional volumes instead.
 * Kafka 4.0 and newer is using Log4j2 for logging instead of Reload4j/Log4j1.
   If you have any custom logging configuration, you might need to update it during the upgrade to Kafka 4.0.
+* KafkaConnect now uses PEM files instead of P12/JKS for keystore and truststore. 
```
```suggestion
* KafkaConnect now uses PEM files instead of PKCS12/JKS for keystore and truststore. 
```

ppatierno commented on api/src/main/java/io/strimzi/api/kafka/model/connect/KafkaConnectResources.java:
```diff
@@ -117,6 +117,17 @@ public static String initContainerClusterRoleBindingName(String clusterName, Str
+     *
+     * @return The name of the init container's cluster role binding.
+     */
+    public static String connectRoleBindingName(String componentName) {
```
+1

tinaselenge replied:
so it would be named as <clusterName>-role rather than <clusterName>-connect-role or <clusterName>-mirrormaker-role?

ppatierno replied:
should we follow a rule like `initContainerClusterRoleBindingName` method?

tinaselenge replied:
That would be something like `strimzi-<namespace>-<clusterName>-connect`? 

I initially followed the format of topic operator and entity operator role binding names. Cluster name is passed and component name is hard-coded to it (e.g. `<clusterName>-entity-topic-operator-role`). I was trying to do it without hard coding `connect` since component name contains it.  

Maybe as Fede suggested, we just pass the cluster name and then do the same as entity/topic operator does for consistency (e.g. `<clusterName>-connect-role`). What do you think?

ppatierno commented on api/src/main/java/io/strimzi/api/kafka/model/connect/KafkaConnectResources.java:
```diff
@@ -117,6 +117,17 @@ public static String initContainerClusterRoleBindingName(String clusterName, Str
+     *
+     * @param componentName  Name of the Kafka Connect component
+     *
+     * @return The name of the init container's cluster role binding.
```
init container? ;-)

tinaselenge replied:
Good catch :D , left from copy paste.

scholzj commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectCluster.java:
```diff
@@ -818,6 +811,56 @@ public ClusterRoleBinding generateClusterRoleBinding() {
+    public Role generateRole() {
+        List<String> certSecretNames = new ArrayList<>();
+        if (tls != null && tls.getTrustedCertificates() != null && !tls.getTrustedCertificates().isEmpty()) {
+            tls.getTrustedCertificates().stream().forEach(certSecretSource -> certSecretNames.add(certSecretSource.getSecretName()));
```
I still think you would need to copy this into a custom Secret as the config provider will not be able to load from multiple secrets. Or you would need to change the config provider design to support multiple secrets. But as I said, I think Connect does not like it when you yank the Secret from underneath it, so having our own Secret would be IMHO better in any case.

tinaselenge replied:
So you don't think separate config provider templates such as `"ssl.truststore.certificates=${strimzisecrets:namespace/first-secret:*.crt}\\n${strimzisecrets:namespace/second-secret:*.crt}"));` would work? Also my understanding was that after this change, Connect does not need to volume mount these secrets as the Secret Config Provider would directly access them. 

Also taking a step back, in what scenarios user might want KafkaConnect to have multiple trusted certificate secret sources? I was thinking about adding a new system test for this kind of scenario, to make sure KafkaConnect trusts certificates from multiple secret sources, but I am not sure how to set this up. We normally would setup a cluster with a listener that maps to a single certificate and key (via `brokerCertChainAndKey`) and port number. Then KafkaConnect cluster is configured to connect to that port and with the CA certificate that signed this certificate for trusstore. I'm struggling to understand, how we would end up with KafkaConnect having to connect to multiple listeners therefore multiple truststore CA certs from multiple secret sources.  Please let me know if I'm not understanding things correctly.

tinaselenge replied:
What I am trying to now understand is that we really need to deal with multiple secret sources or not.

katheris replied:
> So you don't think separate config provider templates such as "ssl.truststore.certificates=${strimzisecrets:namespace/first-secret:*.crt}\\n${strimzisecrets:namespace/second-secret:*.crt}")); would work?

I don't think this is supported and I expect Kafka would either just ignore the second half or fail to parse it at all.

In terms of needing to trust multiple certificates, one example where this could happen is when the CA key for the Kafka brokers is replaced. Since the Kafka brokers have to be rolled to trust new certificates it's possible to be in a situation where some of the Kafka brokers are presenting end-entity certificates signed by the new CA key, while some are presenting end-entity certificates signed by the old CA still. So in that case Connect would need to trust two CA certs to be able to talk to all the brokers still.

scholzj replied:
Yeah, I never tried this, but I did not think this is something that could work. But it might be worth trying - I think we use config providers inside SASL configurations which is a bit similar.

I tried to reproduce the issues I saw in the past with config providers and deleted resources. But I cannot reproduce it anymore. So maybe Kafka fixed it. If you decide to not copy it, it should be probably carefully tested. But I still think having it in our secret would be safer.

> In terms of needing to trust multiple certificates, one example where this could happen is when the CA key for the Kafka brokers is replaced. Since the Kafka brokers have to be rolled to trust new certificates it's possible to be in a situation where some of the Kafka brokers are presenting end-entity certificates signed by the new CA key, while some are presenting end-entity certificates signed by the old CA still. So in that case Connect would need to trust two CA certs to be able to talk to all the brokers still.

Keep in mind this is Connect. So this might be connecting even to non-Strimzi based clusters with many different setups. So it might not be just about how Strimzi updates the CAs etc.

tinaselenge replied:
Thank you both for sharing your thoughts on this. I have tried but it doesn't seem to work with multiple config provider templates. Based on this discussion, I went back to previous change that creates an internal secret for the truststore certificates.

ppatierno commented on api/src/main/java/io/strimzi/api/kafka/model/connect/KafkaConnectResources.java:
```diff
@@ -117,6 +117,30 @@ public static String initContainerClusterRoleBindingName(String clusterName, Str
+     */
+    public static String internalTlsCertsSecretName(String clusterName) {
+        return componentName(clusterName) + "-tls-certs";
+    }
```
just having TLS in both method and Secret name seems too generic. Would it be better to have also "trusted" somewhere? i.e. `internalTlsTrustedCertsSecretName` and "-tls-trusted-certs" suffix?

ppatierno reviewed (commented):
@tinaselenge can you take a look at the failed tests within the build?

tinaselenge commented:
Thanks @ppatierno. The build had failed due to timeout on an unrelated test but after addressing your comment, the build is green.

Can someone please kick off the regression tests on this PR? Thanks!

im-konge commented:
/azp run regression

ppatierno commented on docker-images/kafka-based/kafka/scripts/kafka_connect_run.sh:
```diff
@@ -6,16 +6,9 @@ set +x
-
 # Generate and print the config file
 echo "Starting Kafka Connect with configuration:"
 tee /tmp/strimzi-connect.properties < "/opt/kafka/custom-config/kafka-connect.properties" | sed -e 's/sasl.jaas.config=.*/sasl.jaas.config=[hidden]/g' -e 's/password=.*/password=[hidden]/g'
```
I guess we don't need the `-e 's/password=.*/password=[hidden]/g'` part of this because we don't have PKCS12 stores anymore and so no passoword.

tinaselenge commented:
I created an issue for this PR to discuss some of the points that would also apply when making the other operands use PEM files as well. I thought it would be easier to discuss in an issue than the PR.

ppatierno commented:
/azp run regression

ppatierno reviewed (dismissed):
LGTM. Thanks Tina!

ppatierno commented:
@tinaselenge sorry If I approved but didn't notice your comment on https://github.com/strimzi/strimzi-kafka-operator/issues/11294 so I guess we need more discussion.

tinaselenge commented:
@ppatierno @katheris @fvaleri I have made a change to create an internal secret for trusted certificates for OAuth server when OAuth is used. Can you please review the new change when you have time? Thank you.

Can someone also please kick off the regression tests? Thanks!

ppatierno commented:
/azp run regression

tinaselenge commented:
I'm having an issue with oauth truststore configuration. It seems like Jaas config for the oauth, doesn't like new lines of PEM format certificates. It wants everything inline I think:
```
Caused by: java.lang.IllegalArgumentException: Value not specified for key 'MIIDZTCCAk2gAwIBAgIUYB3lcFeRQUakgAHV6ioDogGuHmAwDQYJKoZIhvcNAQEL' in JAAS config
```
I also hit different issue with kafka-client 4.0 jar:
```
2025-04-08 23:14:10 DEBUG [main] CompositeStrategy:56 - Unable to load modern Subject methods, relying only on legacy methods
java.lang.NoSuchMethodException: javax.security.auth.Subject.current()
	at java.base/java.lang.Class.getDeclaredMethod(Class.java:2675) ~[?:?]
	at org.apache.kafka.common.internals.ModernStrategy.<init>(ModernStrategy.java:43) ~[kafka-clients-4.0.0.jar:?]
	at org.apache.kafka.common.internals.CompositeStrategy.<init>(CompositeStrategy.java:51) [kafka-clients-4.0.0.jar:?]
	at org.apache.kafka.common.internals.CompositeStrategy.<clinit>(CompositeStrategy.java:39) [kafka-clients-4.0.0.jar:?]
	at org.apache.kafka.common.internals.SecurityManagerCompatibility.get(SecurityManagerCompatibility.java:38) [kafka-clients-4.0.0.jar:?]
	at org.apache.kafka.connect.runtime.isolation.ClassLoaderFactory.newDelegatingClassLoader(ClassLoaderFactory.java:30) [connect-runtime-4.0.0.jar:?]
```
Looks it might Java version incompatibility. 

So I'm looking into how to solve this issue. Maybe oauth.ssl.truststore.location in Oauth Jaas config is less complex way to configure it, but we need to combine the trusted certificates into a single file and volume mount it. I don't know if this is too complex, but we could copy oauth trusted cert secrets into an internal secret under a single key and volume mount this internal secret.

ppatierno commented:
/azp run regression

fvaleri commented on systemtest/src/test/java/io/strimzi/systemtest/connect/ConnectST.java:
```diff
@@ -555,17 +555,27 @@ void testSecretsWithKafkaConnectWithTlsAndTlsClientAuthentication() {
+        final String encodedCertMock = Base64.getEncoder().encodeToString("certificates".getBytes());
+
+        Secret secondCertSecret = new SecretBuilder()
+                .withNewMetadata()
+                .withName("my-secret")
+                .withNamespace(testStorage.getNamespaceName())
+                .endMetadata()
```
```suggestion
                .withNewMetadata()
                    .withName("my-secret")
                    .withNamespace(testStorage.getNamespaceName())
                .endMetadata()
```

fvaleri reviewed (approved):
Great work @tinaselenge!

LGTM, just left few minor comments.

tinaselenge commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AuthenticationUtils.java:
```diff
@@ -223,7 +267,58 @@ public static void configureClientAuthenticationVolumeMounts(KafkaClientAuthenti
+     * @param mountOAuthSecretVolumes Indicates whether OAuth secret volume mounts will be added to the list
+     * @param oauthSecretsVolumeMount Path where the OAuth secrets would be mounted
+     */
+    public static void configureClientAuthenticationVolumeMounts2(KafkaClientAuthentication authentication, List<VolumeMount> volumeMountList, String tlsVolumeMount, String passwordVolumeMount, String oauthCertsVolumeMount, String oauthCertsSecretName, String volumeNamePrefix, boolean mountOAuthSecretVolumes, String oauthSecretsVolumeMount) {
```
Yes, I can look into improving the name but the idea was to replace the existing method with this one, once KafkaBridge and MirrorMaker2 are refactored to use only PEM format certificates and use the new method instead of the old one.

katheris replied:
If that is the case we should have the new method as `configureClientAuthenticationVolumeMounts` and rename the old one to something like `configurePCS12ClientAuthenticationVolumeMounts` or `configureLegacyClientAuthenticationVolumeMounts`, so when the old methods are removed we don't need to also rename the new ones

katheris replied:
Also as Fede mentioned can you check the methods in this class and remove the ones that are unused currently.

katheris commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CertUtils.java:
```diff
@@ -226,6 +226,20 @@ public static void createTrustedCertificatesVolumes(List<Volume> volumeList, Lis
+     * @param isOpenShift           Indicates whether we run on OpenShift or not
+     * @param secretName            Name of the internal secret for storing trusted certificates for OAuth server
+     */
+    public static void createOauthTrustedCertificatesVolumes(List<Volume> volumeList, boolean isOpenShift, String secretName) {
```
Why do we need this method? It doesn't seem to do anything specific to oauthTrustedCertificates, can't we call `addNewVolume(List<Volume> volumeList, String volumeNamePrefix, String secretName, boolean isOpenShift)` passing an empty string for `volumeNamePrefix`?

katheris commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CertUtils.java:
```diff
@@ -304,6 +318,21 @@ private static String trustedCertificateVolumeName(CertSecretSource certSecretSo
+     * @param secretName            Prefix used for the volume names to distinguish when the same Secret is mounted for
+     *                              different purposes (e.g. different cluster connections in MM2)
+     */
+    public static void createOauthTrustedCertificatesVolumeMounts(List<VolumeMount> volumeMountList, String tlsVolumeMountPath, String secretName) {
```
Similarly to above this isn't an oauth specific method, so maybe it should be called `addNewVolumeMount`?

katheris reviewed (commented):
The changes broadly look good to me, but definitely need some cleanup of naming in the AuthenticationUtils and CertUtils classes. I also added some suggestions to simplify the code in the AbstractConnectOperator class.

tinaselenge commented:
@katheris @fvaleri thank you for reviewing the PR. I have addressed your comments now :) .

katheris commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/CertUtils.java:
```diff
@@ -304,6 +304,7 @@ private static String trustedCertificateVolumeName(CertSecretSource certSecretSo
         return prefix != null ? prefix + '-' + certSecretSource.getSecretName() : certSecretSource.getSecretName();
     }
 
+
```
You don't need this new line

katheris reviewed (approved):
One tiny nit, but otherwise looks good to me

tinaselenge commented:
thank you Kate!

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AuthenticationUtils.java:
```diff
@@ -121,26 +121,57 @@ private static void checkValueGreaterOrEqualZero(ArrayList<String> errors, Strin
      */
-    public static void configureClientAuthenticationVolumes(KafkaClientAuthentication authentication, List<Volume> volumeList, String oauthVolumeNamePrefix, boolean isOpenShift)   {
-        configureClientAuthenticationVolumes(authentication, volumeList, oauthVolumeNamePrefix, isOpenShift, "", false);
+    public static void configurePKCSClientAuthenticationVolumes(KafkaClientAuthentication authentication, List<Volume> volumeList, String oauthVolumeNamePrefix, boolean isOpenShift, String volumeNamePrefix, boolean createOAuthSecretVolumes)   {
```
`PKCS` is a broader group of standards, which one are you referring to within this name? Maybe the PKCS12 as referring to the p12 store file? But aren't we moving away from it with this PR? I am confused about the method name then or there is something I am missing here :-(

tinaselenge replied:
I will change to PKCS12 as suggested. We are keeping this method, as it's still used by other resources such as Bridge and MirrorMaker as they haven't yet migrated to using PEM format certs.

ppatierno replied:
Correct. So yeah just changing to PKCS12 is enough. Thanks!

ppatierno commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AuthenticationUtils.java:
```diff
@@ -186,14 +219,50 @@ private static void addNewVolume(List<Volume> volumeList, String volumeNamePrefi
      */
-    public static void configureClientAuthenticationVolumeMounts(KafkaClientAuthentication authentication, List<VolumeMount> volumeMountList, String tlsVolumeMount, String passwordVolumeMount, String oauthVolumeMount, String oauthVolumeNamePrefix) {
-        configureClientAuthenticationVolumeMounts(authentication, volumeMountList, tlsVolumeMount, passwordVolumeMount, oauthVolumeMount, oauthVolumeNamePrefix, "", false, null);
+    public static void configurePKCSClientAuthenticationVolumeMounts(KafkaClientAuthentication authentication, List<VolumeMount> volumeMountList, String tlsVolumeMount, String passwordVolumeMount, String oauthCertsVolumeMount, String oauthVolumeNamePrefix, String volumeNamePrefix, boolean mountOAuthSecretVolumes, String oauthSecretsVolumeMount) {
```
Ditto as above.

ppatierno commented on api/src/main/java/io/strimzi/api/kafka/model/connect/KafkaConnectTemplate.java:
```diff
@@ -181,6 +183,26 @@ public void setClusterRoleBinding(ResourceTemplate clusterRoleBinding) {
+
+    public void setRoleBinding(ResourceTemplate roleBinding) {
+        this.roleBinding = roleBinding;
+    }
```
Very stupid question but I didn't get why we need this role and roleBinding field in the template (which is supposed to be something written by the users). Aren't we generating role and rolebinding programmatically for accessing Secrets?

ppatierno commented:
/azp run regression

tinaselenge commented:
@katheris @ppatierno do you think we can merge this now?

MatthiasGabriel commented:
After this change, kafka connect seems to ignore the defined certificate or pattern in tls.trustedCertificates[] and just copies all entries which match .crt (which can be less or more entries depending on the defined secrets)
